add Thread.setName and Thread.getName

vrischmann opened the pull request:
This PR adds `Thread.setName` and `Thread.getName`.

It's not complete, I only tested this on linux. I spent a bit of time researching what other platforms do and from what I can tell setting/getting a thread's name is supported on linux, windows (since [Windows 10](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getthreaddescription)), [macos](https://opensource.apple.com/source/libpthread/libpthread-454.80.2/include/pthread/pthread.h.auto.html), [openbsd](https://man.openbsd.org/pthread_get_name_np.3), [freebsd](https://www.freebsd.org/cgi/man.cgi?query=pthread_getname_np&apropos=0&sektion=3&manpath=FreeBSD+12.2-RELEASE+and+Ports&arch=default&format=html), [netbsd](https://man.netbsd.org/pthread_getname_np.3).
Annoyingly since it's non standard there are some slight differences between implementations of `pthread_setname_np`.

Before implementing everything I'd like to get some early feedback on the API.

LemonBoy commented:
API-wise I think it makes sense to add an extra struct parameter to the thread creation function:
```zig
struct CreateThreadOptions = struct {
    thread_name: ?[:0]const u8 = null,
}
```

There's little to no point to setting the name after the thread is created.

LemonBoy commented on lib/std/Thread.zig:
```diff
@@ -76,6 +76,34 @@ pub fn spinLoopHint() void {
 }
 
+pub fn setName(self: Thread, name: [:0]const u8) !void {
+    if (use_pthreads) {
```
Early exit:
```zig
if (!use_pthreads)
    return;
```

And a pure Zig implementation for Linux can be implemented using musl's `pthread_setname_np` as guide.

vrischmann replied:
ok for the early exit. and yeah I planned to follow musl's implementation.

LemonBoy commented on lib/std/Thread.zig:
```diff
@@ -76,6 +76,34 @@ pub fn spinLoopHint() void {
+
+pub fn getName(self: Thread, allocator: *std.mem.Allocator) !?[:0]const u8 {
+    if (use_pthreads) {
+        // The thread name cannot be longer than 16 bytes including the null terminator.
```
> The thread name  is  a  meaningful  C  language
       string, whose length is restricted to 16 characters, including the ter‐
       minating null byte ('\0').

Is this limitation lifted for other platforms?

vrischmann replied:
I don't think so, I wrote this comment before realising `pthread_setname_np` is non standard and looking at the man pages of the BSDs they don't mention a length limitation. 

I'll change the comment.

semarie replied:
FYI, OpenBSD has 32 bytes limit ('\0' included). the string is copied and silenciously truncated to fit

LemonBoy commented on lib/std/c.zig:
```diff
@@ -276,6 +276,7 @@ pub extern "c" fn pthread_key_create(key: *pthread_key_t, destructor: ?fn (value
 pub extern "c" fn pthread_key_delete(key: pthread_key_t) c_int;
 pub extern "c" fn pthread_getspecific(key: pthread_key_t) ?*c_void;
 pub extern "c" fn pthread_setspecific(key: pthread_key_t, value: ?*c_void) c_int;
+pub extern "c" fn pthread_getname_np(thread: std.c.pthread_t, name: [*:0]const u8, len: usize) c_int;
```
Non portable definitions go into their own `<platform>/c.zig` file, this is not available when targeting eg. Windows.

vrischmann replied:
ok, I'll move it.

daurnimator commented:
> There's little to no point to setting the name after the thread is created.

It can be useful to name worker threads after what they are currently doing.

This also has implications in schedulers where workloads are not locked to specific threads.

vrischmann commented:
I've implemented linux with or without pthreads, macOS and netbsd/freebsd/openbsd. I've only been able to test linux at the moment, I'll be able to test freebsd and probably netbsd.

I'm still working on the windows implementation.

I'll make the API change last.

semarie commented on lib/std/c/openbsd.zig:
```diff
@@ -45,3 +45,6 @@ pub extern "c" fn posix_memalign(memptr: *?*c_void, alignment: usize, size: usiz
 pub extern "c" fn unveil(path: ?[*:0]const u8, permissions: ?[*:0]const u8) c_int;
+
+pub extern "c" fn pthread_set_name_np(thread: std.c.pthread_t, name: [*:0]const u8) c_int;
+pub extern "c" fn pthread_get_name_np(thread: std.c.pthread_t, name: [*:0]u8, len: usize) c_int;
```
these functions are returning `void` on OpenBSD: you could see the [man page](https://man.openbsd.org/pthread_set_name_np).

semarie commented on lib/std/Thread.zig:
```diff
@@ -76,6 +76,121 @@ pub fn spinLoopHint() void {
+            return null;
+        },
+        .macos, .ios, .watchos, .tvos, .netbsd, .freebsd, .openbsd => if (use_pthreads) {
+            const err = c.pthread_getname_np(self.data.handle, buffer.ptr, maxNameLen);
```
`void` return for OpenBSD.

semarie commented on lib/std/Thread.zig:
```diff
@@ -76,6 +76,121 @@ pub fn spinLoopHint() void {
+        },
+        .freebsd, .openbsd => if (use_pthreads) {
+            // Use pthread_set_name_np for FreeBSD because the alias pthread_setname_np is FreeBSD 12.2+ only.
+            const err = c.pthread_set_name_np(self.data.handle, name.ptr);
```
`void` return for OpenBSD

vrischmann replied:
Thanks, these will be fixed.

semarie reviewed (changes requested):
The return type for functions for OpenBSD should be `void`. It will make it different from `.freebsd` for setting the name, and different from others for getting it.

LemonBoy commented on lib/std/Thread.zig:
```diff
@@ -76,6 +76,121 @@ pub fn spinLoopHint() void {
+
+pub fn getName(self: Thread, allocator: *std.mem.Allocator) !?[:0]const u8 {
+    // musl doesn't provide pthread_getname_np and there's no way to retrieve the thread id of an arbitrary thread.
+    if (use_pthreads and self.data.handle != getCurrentId() and comptime std.Target.current.abi.isMusl()) {
```
Nope, let this case use the `prctl`/`proc` code path.

It seems that [musl](https://www.openwall.com/lists/musl/2019/07/17/3) needs a bit of prodding on this front.

vrischmann replied:
I'm not sure I understand. Without this check, it would execute the `proc` reading code but I don't have access to the thread id of an arbitrary `pthread_t` so that won't work.

LemonBoy replied:
Welp, I forgot the `pthread` structure type is opaque. You'd need to dig into the private definition in `pthread_impl.h` but that's pretty ugly... let's return an error instead (`Unsupported` as done for the darwin case maybe?)

vrischmann replied:
Ok, I'll change it to return an error.

LemonBoy commented on lib/std/Thread.zig:
```diff
@@ -76,6 +76,121 @@ pub fn spinLoopHint() void {
+        return null;
+    }
+
+    var buffer = try allocator.allocWithOptions(u8, maxNameLen - 1, null, 0);
```
What if `maxNameLen` is zero?
Wouldn't it make more sense to have `maxNameLen` exclude the tail zero?

vrischmann replied:
My thinking was it would be less surprising for a reader to see the maxNameLen with the same value as in the different man pages, but I'm not particularly attached to that. I'll change it.

LemonBoy replied:
From the Zig user POV the interesting part is the number of characters one can use, the presence of the trailing zero is an implementation detail that leaks to the user.
Moreover you get an extra byte when allocating a buffer using `var buf: [maxNameLen:0]`, dropping the trailing zero is IMO justified.

One more idea, what about a pure Zig API where the user passes in a `[]const u8` slice? Zero-terminating the string on the fly here is pretty cheap and would allow for a more idiomatic API.

vrischmann replied:
(`maxNameLen` changed in the latest commit).

Not sure I get your last point, do you mean something like `fn getName(self: Thread, buffer: []u8) []const u8` ?

LemonBoy replied:
Not quite, I was thinking of something like
```zig
fn getName(self: Thread, allocator: *std.mem.Allocator) !?[]const u8;
fn setName(self: Thread, name: []const u8) !void;
```

where we get rid of the sentinel.

vrischmann replied:
Ah right. Yeah I can do that.

LemonBoy replied:
Keep in mind that Windows `SetThreadDescription` requires a UTF16-LE encoded string so we may need an allocator for `setName` or set an arbitrary upper limit and allocate a static buffer.

LemonBoy commented on lib/std/Thread.zig:
```diff
@@ -76,6 +76,121 @@ pub fn spinLoopHint() void {
+            const err = c.pthread_getname_np(self.data.handle, buffer.ptr, maxNameLen);
+            return switch (err) {
+                0 => std.mem.spanZ(buffer),
+                os.ERANGE => error.NameTooLong,
```
Given the hardcoded upper limit on the length is `maxNameLen` this case can be marked as `unreachable`, no?

LemonBoy commented on lib/std/Thread.zig:
```diff
@@ -76,6 +76,121 @@ pub fn spinLoopHint() void {
+                else => unreachable,
+            };
+        } else if (self.data.handle == getCurrentId()) {
+            _ = try os.prctl(.GET_NAME, .{@ptrToInt(buffer.ptr)});
```
Error checking.

LemonBoy commented on lib/std/Thread.zig:
```diff
@@ -76,6 +76,121 @@ pub fn spinLoopHint() void {
+            // Remove the newline character
+            buffer[read - 1] = 0;
+
+            return std.mem.spanZ(buffer);
```
This can be simplified (and perhaps a check for `read == 0` should be added).

```suggestion
            return buffer[0 .. read - 1];
```

vrischmann replied:
This doesn't work:
```
./lib/std/Thread.zig:186:26: error: expected type '[:0]const u8', found '[]u8'
            return buffer[0 .. read - 1];
                         ^
./lib/std/Thread.zig:186:26: note: destination pointer requires a terminating '0' sentinel
            return buffer[0 .. read - 1];
```

I can do this however:
```
            buffer[read - 1] = 0;

            return buffer[0 .. read - 1 :0];
```
which might be better than using `spanZ`.

LemonBoy replied:
Indeed, I missed the `:0`.
But do we really want to return a zero-terminated slice instead of a "classic" one?

vrischmann replied:
Yeah you're right, we could just return a `[]const u8`.

vrischmann commented:
I've addressed most of the comments in the latest commit (minus those still in discussion).

I successfully tested this on FreeBSD, however I can't test on NetBSD or OpenBSD because of compilation errors somewhere else in the stdlib.

vrischmann commented:
I've added the windows implementation. 

One thing I'm not sure about is the utf8 -> utf16 conversion, right now I'm doing this:
```
           var name_buf_w: [maxNameLen:0]u16 = undefined;
           const length = try std.unicode.utf8ToUtf16Le(&name_buf_w, name);
```
The utf8 `name` can't be bigger than `maxNameLen` so I was assuming a `[maxNameLen:0]u16` is always enough to fit the utf16 data but I'm not sure if that's correct, I'm not super familiar with UTF16.

LemonBoy commented on lib/std/Thread.zig:
```diff
@@ -76,6 +76,188 @@ pub fn spinLoopHint() void {
+
+            try os.windows.SetThreadDescription(
+                self.data.handle,
+                std.meta.cast(os.windows.LPWSTR, name_buf_w[0..length :0]),
```
Nit, `meta.cast` is not supposed to be used outside of `translate-c`'d code. A `@ptrCast` is enough here.

vrischmann replied:
Ok, I never realized that.

MasterQ32 replied:
Imho, a good ol `&name_buf_w` should do the job here, no need to slice (we know it's NUL terminated anyways) and the function seems to accept a `[*:0]const u16`.

LemonBoy commented on lib/std/Thread.zig:
```diff
@@ -76,6 +76,188 @@ pub fn spinLoopHint() void {
+                else => return os.unexpectedErrno(err),
+            };
+        },
+        .freebsd => if (use_pthreads) {
```
You can merge this with the `openbsd` case below.

semarie replied:
it was splitted because freebsd function is returning `int` and openbsd function is returning `void` (previously the return value was checked).

but as it, freebsd case doesn't check for `pthread_get_name_np` return value.

vrischmann replied:
I opted to keep them split right now mainly to add the FreeBSD comment, but I can also add the comment in the `freebsd, openbsd` case.

LemonBoy commented on lib/std/Thread.zig:
```diff
@@ -572,6 +754,82 @@ pub fn getCurrentThreadId() u64 {
+
+    comptime const testCases = &[_][]const u8{
+        "mythread",
+        max_length_name,
```
Won't `"b" ** maxNameLen` work?

LemonBoy commented on lib/std/Thread.zig:
```diff
@@ -572,6 +754,82 @@ pub fn getCurrentThreadId() u64 {
+    std.testing.expectError(error.Unsupported, res2);
+}
+
+test "Thread: setName, getName" {
```
No need to prefix the test names.

LemonBoy commented on lib/std/Thread.zig:
```diff
@@ -572,6 +754,82 @@ pub fn getCurrentThreadId() u64 {
+        done: std.atomic.Bool = std.atomic.Bool.init(false),
+
+        pub fn run(ctx: *@This()) !void {
+            while (!ctx.done.load(.SeqCst)) {
```
I'd suggest to perform the `getName` and `setName` within the thread itself, otherwise you're not testing anything on darwin systems.

LemonBoy commented on lib/std/Thread.zig:
```diff
@@ -572,6 +754,82 @@ pub fn getCurrentThreadId() u64 {
+            const res = thread.setName("foobar");
+            std.testing.expectError(error.Unsupported, res);
+        },
+        .windows => if (std.Target.current.os.isAtLeast(.windows, .win10_rs1)) |res| {
```
`isAtLeast` returns `null` if a runtime-check is required, I'd simply run `testThreadName` and catch `unsupported`, in that case you can return `SkipZigTest` as that's not an error condition.

MasterQ32 commented on lib/std/Thread.zig:
```diff
@@ -76,6 +76,188 @@ pub fn spinLoopHint() void {
 
+const maxNameLen = switch (std.Target.current.os.tag) {
+    .linux => 15,
+    .windows => 31,
```
Where does this limit come from? I've read the documentation about SetThreadDescription and it doesn't mention a 31 (or 32) as a maximum length

vrischmann replied:
It's arbitrary, I chose something which seems appropriate for a thread name.

PhaseMage commented:
> One thing I'm not sure about is the utf8 -> utf16 conversion, right now I'm doing this:
> 
> ```
>            var name_buf_w: [maxNameLen:0]u16 = undefined;
>            const length = try std.unicode.utf8ToUtf16Le(&name_buf_w, name);
> ```
> 
> The utf8 `name` can't be bigger than `maxNameLen` so I was assuming a `[maxNameLen:0]u16` is always enough to fit the utf16 data but I'm not sure if that's correct, I'm not super familiar with UTF16.

This is correct. A string of 'n' UTF16 code units can always hold 'n' UTF8 code units. 

The normal issue is only the other way around, where one UTF16 code unit might require multiple UTF8 code units of storage. For ASCII code points, this isn't relevant. But for larger code points (eg. Emoji), this can be important. I suspect that Windows supports a maximum number of UTF16 code units. Since you allocate to implement getName(), this works fine. However, limiting maxNameLen to 31 UTF8 code units artificially limits the number of code points you can give to setName() (say, a string of Emoji). But using Emoji in thread names is probably a bad idea anyway since debuggers might not support it well :-)

vrischmann commented:
> > One thing I'm not sure about is the utf8 -> utf16 conversion, right now I'm doing this:
> > ```
> >            var name_buf_w: [maxNameLen:0]u16 = undefined;
> >            const length = try std.unicode.utf8ToUtf16Le(&name_buf_w, name);
> > ```
> > 
> > 
> > The utf8 `name` can't be bigger than `maxNameLen` so I was assuming a `[maxNameLen:0]u16` is always enough to fit the utf16 data but I'm not sure if that's correct, I'm not super familiar with UTF16.
> 
> This is correct. A string of 'n' UTF16 code units can always hold 'n' UTF8 code units.
> 
> The normal issue is only the other way around, where one UTF16 code unit might require multiple UTF8 code units of storage. For ASCII code points, this isn't relevant. But for larger code points (eg. Emoji), this can be important. I suspect that Windows supports a maximum number of UTF16 code units. Since you allocate to implement getName(), this works fine. However, limiting maxNameLen to 31 UTF8 code units artificially limits the number of code points you can give to setName() (say, a string of Emoji). But using Emoji in thread names is probably a bad idea anyway since debuggers might not support it well :-)

Thanks. I actually tested emojis in the thread name, Visual Studio 2019 supports them at least. 
I agree that a `maxNameLen` of 31 is arbitrary but in my opinion it should be enough, and it's in line with what other platforms do.
I could try and see if the call to `SetThreadDescription` fails with a bigger maxNameLen and get a better idea where the limit is, because I couldn't find it in the documentation.

andrewrk commented on lib/std/Thread.zig:
```diff
@@ -70,6 +70,181 @@ else switch (std.Target.current.os.tag) {
+    else => 0,
+};
+
+pub fn setName(self: Thread, name: []const u8) !void {
```
This has different errors across different targets and so it must have an explicit error set (e.g. so that "Unsupported" shows up even on targets where it is always supported).

vrischmann replied:
Addressed in the latest commit

andrewrk commented on lib/std/Thread.zig:
```diff
@@ -70,6 +70,181 @@ else switch (std.Target.current.os.tag) {
+    }
+}
+
+pub fn getName(self: Thread, allocator: *std.mem.Allocator) !?[]const u8 {
```
I suggest to take an array buffer of `maxNameLen` length, and then return a slice into that buffer. This will avoid the need for an allocator.

If you want you can then provide a higher level API, `getNameAlloc`, implemented using the lower level API.

vrischmann replied:
Addressed in the latest commit

andrewrk reviewed (changes requested):
Thanks for your patience; I know it took a long time to get a review from me.

I have a couple requests and then I think this looks good to merge.

Can you please rebase against master branch and fix the conflicts in addition to the requested changes?

andrewrk commented on lib/std/Thread.zig:
```diff
@@ -38,6 +38,180 @@ else
 
 impl: Impl,
 
+pub const maxNameLen = switch (std.Target.current.os.tag) {
```
This is public so let's make it conform to the [style guide](https://ziglang.org/documentation/master/#Style-Guide):

```suggestion
pub const max_name_len = switch (std.Target.current.os.tag) {
```

andrewrk commented on lib/std/Thread.zig:
```diff
@@ -38,6 +38,180 @@ else
+    }
+}
+
+pub fn getName(self: Thread, buffer_ptr: *[maxNameLen:0]u8) !?[]const u8 {
```
This also needs an explicit error set to make it the same across targets.

andrewrk reviewed (changes requested):
Thanks for the fixups! Found 2 more things - sorry for not finding them last time, but at least now I am paying close attention, and will give this a prompt merge :)

vrischmann commented:
Ok, addressed your last feedback.

I had to change some logic for musl since the type returned by `getCurrentId` changed but it's working now.
I've only been able to test linux/windows, I can test freebsd later.
I don't have a mac to test on macOS unfortunately.

vrischmann commented:
Pushed the last fixes for macos, freebsd/netbsd.
